internal/pkg/action/operator_list_custom_resources.go

```diff
@@ -0,0 +1,113 @@
+	Unzip(ctx context.Context, operator *v2alpha1.Operator) ([]v1alpha1.CRDDescription, error)
+}
+
+// OperatorListCustomResources knows how to find and list custom resources given a package name and namespace.
+type OperatorListCustomResources struct {
+	config  *Configuration
+	ListAll bool
+	PackageName string
+	Namespace string
+}
+
+func NewOperatorListCustomResources(cfg *Configuration) *OperatorListCustomResources {
+	return &OperatorListCustomResources{
+		config: cfg,
+		Namespace: cfg.Namespace,
+	}
+}
+
+// FindOperator finds an operator object on-cluster provided a package and namespace. Returns a copy of the operator.
+func (o *OperatorListCustomResources) FindOperator(ctx context.Context) (v2alpha1.Operator, error) {
+	opKey := types.NamespacedName{
+		Name:      o.PackageName,
+		Namespace: o.Namespace,
+	}
+	operator := v2alpha1.Operator{}
+	err := o.config.Client.Get(ctx, opKey, &operator)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return v2alpha1.Operator{}, fmt.Errorf("could not find operator for package %s in namespace %s", o.PackageName, o.Namespace)
+		}
+		return v2alpha1.Operator{}, err
+	}
+	return operator, nil
+}
+
+// Unzip finds the CSV referenced by the provided operator and then inspects the spec.customresourcedefinitions.owned
+// section of the CSV to return a list of APIs that are owned by the CSV.
+func (o *OperatorListCustomResources) Unzip(ctx context.Context, operator *v2alpha1.Operator) ([]v1alpha1.CRDDescription, error) {
+	csv := v1alpha1.ClusterServiceVersion{}
+	csvKey := types.NamespacedName{}
+
+	if operator.Status.Components == nil {
+		return nil, fmt.Errorf("could find underlying components for operator %s", operator.Name)
+	}
+	for _, resource := range operator.Status.Components.Refs {
+		if resource.Kind == v1alpha1.ClusterServiceVersionKind {
+			csvKey.Name = resource.Name
+			csvKey.Namespace = resource.Namespace
+			break
+		}
+		return nil, fmt.Errorf("could find underlying CSV for operator %s", operator.Name)
+	}
+
+	err := o.config.Client.Get(ctx, csvKey, &csv)
+	if err != nil {
+		return nil, fmt.Errorf("could not get %s CSV on cluster: %s", csvKey.String(), err)
+	}
+
+	// check if owned CRDs are defined on the csv first
+	if reflect.DeepEqual(csv.Spec.CustomResourceDefinitions, v1alpha1.CustomResourceDefinitions{}) || len(csv.Spec.CustomResourceDefinitions.Owned) == 0 {
+		return nil, fmt.Errorf("no owned CustomResourceDefinitions specified on CSV %s, no custom resources to display", csvKey.String())
+	}
+
+	return csv.Spec.CustomResourceDefinitions.Owned, nil
+}
+
+// List takes in a list of CRDs and finds the associated CRs metadata on-cluster using the client-go meta client.
+// Note: List() can return a potentially unbounded list that callers may need to paginate.
+func (o *OperatorListCustomResources) List(ctx context.Context, crds []v1alpha1.CRDDescription) (result metav1.PartialObjectMetadataList, resErr error) {
```

Comment:
Any reason for the named return values? I typically try to avoid using them unless there's a specific need because it makes `return` statements harder to grok.

This problem is also mentioned in the Tour of Go: https://tour.golang.org/basics/7

Reply:
I generally also tend not to use named return values, although in this case I thought it could be helpful to have a named `Results` return variable. Totally fine with me to refactor back to standard return values.

Reply:
Yeah I figured that was the reason. I used to do that with functions that just appended to the return value in a loop, but I've tripped myself up often enough on naked returns that I just put `var out []ReturnType` as the first line in those functions and call it a day.

The only time I think I really use named return values is when I need a `defer`-ed function to modify the values on the way out.